src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,184 @@
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addFirst(T value) {
+        Node<T> h = head;
+        Node<T> newNode = new Node<>(null, value, h);
+        head = newNode;
+        if (h == null) {
+            tail = newNode;
+        } else {
+            h.prev = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        Node<T> t = tail;
+        Node<T> newNode = new Node<>(t, value, null);
+        tail = newNode;
+        if (t == null) {
+            head = newNode;
+        } else {
+            t.next = newNode;
+        }
+        size++;
+    }
+
+    private void addBefore(T value, Node<T> node) {
+        Node<T> act = node.prev;
+        Node<T> actual = new Node<>(act, value, node);
+        node.prev = actual;
+        if (act == null) {
+            head = actual;
+        } else {
+            act.next = actual;
+        }
+        size++;
+    }
+
+    private Node<T> findNode(int index) {
+        Node<T> search;
+        if (index < (size / 2)) {
+            search = head;
+            for (int i = 0; i < index; i++) {
+                search = search.next;
+            }
+        } else {
+            search = tail;
+            for (int i = size - 1; i > index; i--) {
+                search = search.prev;
+            }
+        }
+        return search;
+    }
+
+    private T unlink(Node<T> element) {
+        final T unlinkValue = element.value;
+        Node<T> next = element.next;
+        Node<T> prev = element.prev;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            element.prev = null;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            element.next = null;
+        }
+
+        element.value = null;
+        size--;
+        return unlinkValue;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Non-existent position");
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("index out of bounds");
+        }
+    }
```

Comment:
in my opinion, these two methods can be combined into one (yours is checkElementIndex) with the message "Index Out Of Bounds!", by adding size to the signature and in the add(T value, int index) method, call checkIndex(index, size + 1);